parser: parse Unicode escape sequences in keys

belochub opened the pull request:
Also, add support for Unicode surrogate pairs and improve key parsing performance by avoiding copying when possible.
Refs: https://github.com/metarhia/jstp/issues/152

aqrln commented:
@belochub can you please rebase it? I think removing the old `package-lock.json`, resolving the conflict in `package.json` and regenerating the package-lock should be just fine.

nechaido commented:
@belochub can you, please, rebase this PR on master?

belochub commented:
@aqrln, @lundibundi, ping.

aqrln commented on src/jsrs_parser.cc:
```diff
@@ -554,6 +555,57 @@ MaybeLocal<Value> ParseString(Isolate*    isolate,
+
+  if (isxdigit(str[0])) {
+    result = ReadHexNumber(str, 4, ok);
+    *size = 4;
```
Should it check whether `*ok` is true before changing `*size`?

aqrln commented on src/jsrs_parser.cc:
```diff
@@ -554,6 +555,57 @@ MaybeLocal<Value> ParseString(Isolate*    isolate,
+        return 0xFFFD;
+      }
+    }
+    result = ReadHexNumber(str + 1, hex_size, ok);
```
It should check whether `result` is a valid code point (i.e., something like `\u{1234123412341234}` must be a syntax error).

belochub replied:
Yeah, I looked into it and decided to reimplement `ReadHexNumber` overall, because it has some other problems.

belochub commented:
@aqrln, I've updated `ReadHexNumber` function to make it safer and added tests for it, PTAL.
Also, this probably fixed parsing of strings like this: `'\u00001'` with the digits after the escape sequence, this case was incorrectly considered an error before.

aqrln commented:
Landed in bef3572c2967d4da668d92e7d944efd58ca3f15b.
